fastcrypto/src/bls12381.rs

```diff
@@ -206,23 +206,35 @@ impl VerifyingKey for BLS12381PublicKey {
         pks: &[Self],
         sigs: &[Self::Sig],
     ) -> Result<(), eyre::Report> {
-        let num_sigs = sigs.len();
+        // TODO: fix this, the identical message opens up a rogue key attack
+        let msgs_refs = (0..sigs.len()).map(|_| msg).collect::<Vec<_>>();
+        Self::verify_batch_empty_fail_different_msg(&msgs_refs, pks, sigs)
+    }
+
+    fn verify_batch_empty_fail_different_msg<'a, M>(
+        msgs: &[M],
+        pks: &[Self],
+        sigs: &[Self::Sig],
+    ) -> Result<(), eyre::Report>
+    where
+        M: Borrow<[u8]> + 'a,
+    {
         if sigs.is_empty() {
             return Err(eyre!(
                 "Critical Error! This behaviour can signal something dangerous, and \
             that someone may be trying to bypass signature verification through providing empty \
             batches."
             ));
         }
-        if sigs.len() != pks.len() {
+        if sigs.len() != pks.len() || msgs.len() != pks.len() {
             return Err(eyre!(
-                "Mismatch between number of signatures and public keys provided"
+                "Mismatch between number of messages, signatures and public keys provided"
             ));
         }
-        let mut rands: Vec<blst_scalar> = Vec::with_capacity(num_sigs);
+        let mut rands: Vec<blst_scalar> = Vec::with_capacity(sigs.len());
         let mut rng = OsRng;
 
-        for _ in 0..num_sigs {
+        for _ in 0..sigs.len() {
```

Comment:
Unrelated to this PR, but we need a TODO fix here as the rand scalars should be at least 128bits (thus we should set vals[1] as well). Ideally I'd create a private function get_128bit_scalar and invoke that.